File: python/place_order.py
"""Place a perp limit order (BTC, IOC, ~$10 notional)."""

import json
from client import exchange, sign_hash, get_mid

COIN = "BTC"

mid = get_mid(COIN)
if mid == 0:
    print(f"Could not fetch {COIN} mid price")
    raise SystemExit(1)

sz = round(11.0 / mid, 5)
buy_px = int(mid * 1.03)

print(f"{COIN} mid: ${mid:,.2f}")
print(f"BUY {sz} @ {buy_px} (IOC, ~${sz * mid:.2f} notional)")

res = exchange({
    "action": {
        "type": "order",
        "orders": [{
            "asset": COIN,
            "side": "buy",
            "price": str(buy_px),
            "size": str(sz),
            "tif": "ioc",
        }],
    },
})
sig = sign_hash(res["hash"])

result = exchange({
    "action": res["action"],
    "nonce": res["nonce"],
    "signature": sig,
})

print(json.dumps(result["exchangeResponse"], indent=2))

File: typescript/src/placeOrder.ts
import { exchange, signHash, getMid } from "./client";

const COIN = "BTC";

async function main() {
  const mid = await getMid(COIN);
  if (mid === 0) {
    console.error(`Could not fetch ${COIN} mid price`);
    process.exit(1);
  }

  const sz = (11.0 / mid).toFixed(5);
  const buyPx = Math.floor(mid * 1.03).toString();

  console.log(`${COIN} mid: $${mid.toLocaleString()}`);
  console.log(`BUY ${sz} @ ${buyPx} (IOC, ~$${(parseFloat(sz) * mid).toFixed(2)} notional)`);

  const res = await exchange({
    action: {
      type: "order",
      orders: [{ asset: COIN, side: "buy", price: buyPx, size: sz, tif: "ioc" }],
    },
  });
  const sig = await signHash(res.hash);

  const result = await exchange({
    action: res.action,
    nonce: res.nonce,
    signature: sig,
  });

  console.log(JSON.stringify(result.exchangeResponse, null, 2));
}

main();
